ds3-autogen-go/src/main/kotlin/com/spectralogic/ds3autogen/go/generators/client/command/BaseCommandGenerator.kt

```diff
@@ -0,0 +1,149 @@
+import com.spectralogic.ds3autogen.go.utils.toGoType
+import com.spectralogic.ds3autogen.utils.ClientGeneratorUtil
+import com.spectralogic.ds3autogen.utils.ConverterUtil.isEmpty
+import com.spectralogic.ds3autogen.utils.Helper
+import com.spectralogic.ds3autogen.utils.collections.GuavaCollectors
+import java.util.*
+
+open class BaseCommandGenerator : CommandModelGenerator<Command>, CommandGeneratorUtil {
+
+    override fun generate(ds3Request: Ds3Request): Command {
+        return Command(
+                ClientGeneratorUtil.toCommandName(ds3Request.name),
+                toRequestBuildLines(ds3Request))
+    }
+
+    /**
+     * Creates the list of request handler build lines required to construct the
+     * http request using a builder within the client. This excludes the request
+     * builder initialization (first line) and the build command (last line).
+     */
+    fun toRequestBuildLines(ds3Request: Ds3Request): ImmutableList<RequestBuildLine> {
+        val builder = ImmutableList.builder<RequestBuildLine>()
+        builder.add(HttpVerbBuildLine(ds3Request.httpVerb))
+        builder.add(PathBuildLine(ds3Request.toRequestPath()))
+        builder.addAll(toRequiredQueryParamBuildLines(ds3Request.requiredQueryParams))
+        builder.addAll(toOptionalQueryParamBuildLines(ds3Request.optionalQueryParams))
+
+        // Add optional build lines
+        toOperationBuildLine(ds3Request.operation).ifPresent { line -> builder.add(line) }
+        toReaderBuildLine().ifPresent{ line -> builder.add(line) }
+        toChecksumBuildLine().ifPresent{ line -> builder.add(line) }
+        toHeadersBuildLine().ifPresent { line -> builder.add(line) }
+        return builder.build()
+    }
+
+    /**
+     * Retrieves the request builder lines for adding required query parameters.
+     */
+    override fun toRequiredQueryParamBuildLines(requiredParams: ImmutableList<Ds3Param>?): ImmutableList<RequestBuildLine> {
+        if (isEmpty(requiredParams)) {
+            return ImmutableList.of()
+        }
+
+        return requiredParams!!.stream()
+                .filter { param -> param.name != "Operation" }
+                .map { param -> toRequiredQueryParamBuildLine(param) }
+                .collect(GuavaCollectors.immutableList())
+    }
+
+    /**
+     * Generates the request builder line for adding this required query parameter.
+     */
+    fun toRequiredQueryParamBuildLine(ds3Param: Ds3Param): RequestBuildLine {
+        val goType = toGoType(ds3Param.type, nullable = false)
+        val goName = toGoParamName(ds3Param.name, ds3Param.type)
+        val key = Helper.camelToUnderscore(ds3Param.name)
+        val value = goQueryParamToString(goName, goType)
+        return QueryParamBuildLine(key, value)
+    }
+
+    /**
+     * Retrieves the request builder lines for adding optional query parameters.
+     */
+    override fun toOptionalQueryParamBuildLines(optionalParams: ImmutableList<Ds3Param>?): ImmutableList<RequestBuildLine> {
```

Comment:
Same question about if `optionalParams` can be null.  Really I want to know why it is nullable, and if we can fix this so that it does not accept null.

Reply:
Its in the parsing code module. I can probably update it, but it will touch all modules and should really be a separate PR.